src/main/thrift/parquet.thrift

```diff
@@ -595,6 +607,62 @@ union ColumnOrder {
   1: TypeDefinedOrder TYPE_ORDER;
 }
 
+struct PageLocation {
+/** Offset of the page in the file **/
+  1: required i64 offset
+
+/**
+ * Size of the page, including header. Sum of compressed_page_size and header
+ * length
+ */
+  2: required i32 compressed_page_size
+
+/**
+ * Index within the RowGroup of the first row of the page; this means pages
+ * change on record boundaries (r = 0).
+ */
+  3: required i64 first_row_index
+}
+
+struct OffsetIndex {
+/**
+ * PageLocations, ordered by increasing PageLocation.offset. It is required
+ * that page_locations[i].first_row_index > page_locations[i+1].first_row_index.
+ */
+  1: required list<PageLocation> page_locations
+}
+
+/**
+ * Description for ColumnIndex.
+ * Each <array-field>[i] refers to the page at OffsetIndex.page_locations[i]
+ */
+struct ColumnIndex {
+/**
+ * A list of bools to determine the validity of the corresponding min and max
+ * values. If this is true, the corresponding min and max values in the next
+ * two lists are non-null.
+ */
+  1: required list<bool> null_pages
+
+/**
+ * A list containing the lower bounds for the values of each page. This may be
+ * the actual minimum value found on a page, but can also be a (more compact)
+ * value that does not exist on that page. For ordered columns, these values
+ * must be lower and upper bounds for consecutive pages respectively.
+ */
+  2: required list<binary> min_values
```

Comment:
I like where you're going with the description, but I was initially confused by it. How about giving a concrete example?

It may also be good to note that descending order still stores "min" values here, meaning the values that come first in the ordering.

Reply:
The guarantee for ordered columns is only required if `max_values` is omitted. The reason is that it's hard to come up with a value between one page's max and the next's min when truncating. Imagine I have a page that ends with "aaaaaaaaaaaaaaaaa" and the next one starts with "aaaaaaaaaaaaaaaaaaaab". If we're truncating strings to save space, we don't want to add logic to come up with a value between the two for the min. It's better in that case to truncate both and also write `max_values`.

Should we also clarify that the guarantee is that `x >= lower` and `x <= upper` for all values `x` in a page? I think we should be explicit that bounds could be equal.

Reply:
Also we should clarify where we imply strictly lower vs lower. The current definition of min and max always implies "or equal" which means that the boundary values in min_values here can be in both pages. It is not min included and max excluded as you would expect for boundary values (unless we add the constraint of not starting a new page as long as the same value is repeated). This lowers the value of making max_values optional. I think it general everything is simpler if we keep max_values. Is it really worth making optional?